email/fetcher.go

```diff
@@ -0,0 +1,393 @@
+		if err != nil {
+			logger.Errorf("Failed listing messages, error: %v\n", err)
+		}
+	}()
+
+	// build the map
+	var ids []uint32
+	for msg := range messageChan {
+		ids = append(ids, msg.Uid)
+	}
+	return ids, nil
+}
+
+// getMessagesToFetch returns which messages are not in our database
+//
+// TODO: improve performance, there's no need to do N findOne's
+func (f *Fetcher) getMessagesToFetch(mailbox *imap.MailboxStatus, msgs []uint32) ([]uint32, error) {
+	// convenience variables
+	database := f.staticDatabase
+	logger := f.staticLogger
+
+	// create an array to hold the messages that are missing
+	toFetch := make([]uint32, 0, len(msgs))
+	for _, msgUid := range msgs {
+		uid := buildMessageUID(mailbox, msgUid)
+		email, err := database.FindOne(uid)
+		if err != nil {
+			logger.Errorf("failed to find message '%v', error: %v", msgUid, err)
+			continue
+		}
+
+		// if not append it to the list of ids to fetch
+		if email == nil {
+			toFetch = append(toFetch, msgUid)
+		}
+	}
+	return toFetch, nil
+}
+
+// fetchMessages fetches all messages in the given seq set and persists them in // the database
+func (f *Fetcher) fetchMessages(mailbox *imap.MailboxStatus, toFetch *imap.SeqSet) error {
+	// convenience variables
+	email := f.staticEmailClient
+	logger := f.staticLogger
+
+	messageChan := make(chan *imap.Message)
+	section, err := imap.ParseBodySectionName("BODY[]")
+	if err != nil {
+		return err
+	}
+	done := make(chan error, 1)
+	go func() {
+		done <- email.UidFetch(toFetch, []imap.FetchItem{imap.FetchEnvelope, section.FetchItem()}, messageChan)
+	}()
+
+	toUnsee := new(imap.SeqSet)
+	for msg := range messageChan {
+		// skip messages that have been sent by the abuse scanner itself, since
+		// we reply to the original email those replies are picked up by the
+		// scanner as well
+		if isFromAbuseScanner(msg) {
+			logger.Debugf("skip message from abuse scanner (expected)")
+			err := f.persistSkipMessage(mailbox, msg)
+			if err != nil {
+				logger.Errorf("Failed to persist skip message, error: %v\n", err)
+			}
+			continue
+		}
+
+		// skip messages without body
+		//
+		// TODO: side-effect from UidFetch and can probably be avoided
+		if !hasBody(msg) {
+			logger.Debugf("skip message due to not having a body (expected)")
+			err := f.persistSkipMessage(mailbox, msg)
+			if err != nil {
+				logger.Errorf("Failed to persist skip message, error: %v\n", err)
+			}
+			continue
+		}
+
+		toUnsee.AddNum(msg.Uid)
+		err := f.persistMessage(mailbox, msg, section)
+		if err != nil {
+			logger.Errorf("Failed to persist %v, error: %v\n", msg.Uid, err)
+		}
+	}
+
+	// unsee messages
+	flags := []interface{}{imap.SeenFlag}
+	err = email.UidStore(toUnsee, "-FLAGS.SILENT", flags, nil)
+	if err != nil && !strings.Contains(err.Error(), "Could not parse command") {
+		logger.Debugf("Failed to unsee messages, error: %v\n", err)
+	} else {
+		logger.Debugln("Successfully unseen messages")
+	}
+
+	return <-done
```

Comment:
Initially I thought you were doing this in order to block here until we send to this channel. Then I realised that you are just returning the error value. Can you add a comment to explain that? It's clever but as such it is not obvious.